homeassistant/components/blink/__init__.py

```diff
@@ -84,28 +85,30 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     auth_data = deepcopy(dict(entry.data))
     blink.auth = Auth(auth_data, no_prompt=True, session=session)
     blink.refresh_rate = entry.options.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
+    coordinator = BlinkUpdateCoordinator(hass, blink)
 
     try:
-        await blink.start()
+        await coordinator.api.start()
     except (ClientError, asyncio.TimeoutError) as ex:
         raise ConfigEntryNotReady("Can not connect to host") from ex
 
-    if blink.auth.check_key_required():
+    if coordinator.api.auth.check_key_required():
         _LOGGER.debug("Attempting a reauth flow")
         raise ConfigEntryAuthFailed("Need 2FA for Blink")
 
-    hass.data[DOMAIN][entry.entry_id] = blink
+    hass.data[DOMAIN][entry.entry_id] = coordinator
 
     if not blink.available:
         raise ConfigEntryNotReady
 
     await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
     entry.async_on_unload(entry.add_update_listener(update_listener))
-    await blink.refresh(force=True)
+
+    await coordinator.async_config_entry_first_refresh()
```

Comment:
I notice that you removed the `update_before_add` from the `async_add_entities`. That's usually the way to go since we refresh the coordinator before we forward the entry setups to the platforms. But in this case we refresh the data after we forward the entry setups. Why? Does the coordinator already have data before the first refresh? If so, why do we refresh?

Reply:
I've made this change, but now the coordinator update code is no longer being executed in my tests.
Adding `await coordinator.async_config_entry_first_refresh()` back returns the execution of this code:
https://github.com/home-assistant/core/blob/5dc12410653b68ada53ee9e6bfea54892aefa3d8/homeassistant/components/blink/coordinator.py#L32-L41
Isn't update_before_add supposed to trigger this code?  Clearly I'm confused :)

Reply:
Right, let me give a shot to explaining.

`update_before_add` is a thing that causes core to run `async_update` of the entities before the entities (and their state) is added to HA. This way, the entities have a moment to get their state right before it's being written.

With coordinators, we usually run `async_config_entry_first_refresh` in `__init__.py` to make sure the coordinator has data. And now that I am typing this, I remember that your entities are hooked in on `_async_coordinator_updated` and they are not subscribed at the moment I suggested to add `async_config_entry_first_refresh`.

Reply:
I have seen this setup before but now I am not sure how they work.

Reply:
I just checked it, and I think Airzone is a good example. They have a function `_async_update_attrs` which updates the attributes. This function is called at  `_handle_coordinator_update` and in the constructor of the entity.

Reply:
Thanks for the example - makes more sense now. Sure is easy to get lost when making this type of change (and probably following a bad example too).

Reply:
I think with the code at the point where I started reviewing would go unavailable for a split second until the first coordinator got data. So I am happy that I now also know how to use this.